File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        size_t n = nums.size();

        if(n == 1) return nums[0];
        if(n == 2) return max(nums[0], nums[1]);

        int next = 0, next1 = 0, curr0 = 0;

        for(int i=n-2; i>=0; --i){
            curr0 = max(nums[i]+next1, next);
            next1 = next;
            next = curr0;
        }

        next = 0; next1 = 0;
        int curr1 = 0;
        for(int i=n-1; i>=1; --i){
            curr1 = max(nums[i]+next1, next);
            next1 = next;
            next = curr1;
        }

        return max(curr1, curr0);
    }
};

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if len(nums) <= 3: return max(nums)
        dp = [[-1]*2 for _ in range(len(nums))]
        def rec(i: int, flag: bool) -> int:
            if i >= n or (flag and i == len(nums)-1): return 0
            if dp[i][flag] != -1: return dp[i][flag]
            dp[i][flag] = max(rec(i+1,flag), nums[i]+rec(i+2,flag))
            return dp[i][flag]
        return max(rec(0,True),rec(1,False))